text/0000-succinct-atomic-swaps.md

```diff
@@ -0,0 +1,174 @@
+
+## Atomic nonces
+[atomic-nonces]: #atomic-nonces
+
+In the case of Grin, atomic nonces are just Secp256k1 private keys used to unlock funds on the other chain. Let's say Alice knows atomic nonce `nA`, and Bob knows atomic nonce `nB`.
+
+In the refund transaction, Alice (with locked Grin funds) will reveal her nonce `nA` to allow Bob to recover his locked funds on the other chain.
+
+In the main transaction, Bob and Alice create a multisignature Grin transaction which reveals Bob's nonce `nB` to Alice. Alice can then recover the funds locked on the other chain.
+
+Atomic IDs are identifiers used to derive atomic nonces using the keychain. Atomic IDs are formatted as follows:
+
+```
+0x4 | 'mwatomic' | 0x0 0x0 0x0 0x0 | 32-bit ID
+```
+
+The 32-bit ID can be expanded to 64-bits if necessary, and doesn't need to be random (it can be a simple counter). However, an ID can only be used once.
+
+Currently, output blinding factors and signature nonces use the key ID of the inputs and outputs, which is derived from the current child key path, or next available key path.
+
+Atomic swaps still use the default methods for deriving blinding factors and signature nonces. The atomic IDs are domain separated to help ensure they don't collide with normal blinding factors and signature nonces.
+
+Atomic nonces must only be used **once**. Reuse potentially results in loss of funds, if the other party has the previously used nonce.
+
+Without automated defenses, a malicious party could trick a user into deriving the same nonce (by supplying a duplicate atomic ID). The malicious user could then recover funds on the other chain without completing the atomic swap protocol.
+
+Regular blinding factors and signature nonces use a strictly increasing counter to prevent deriving the same factors more than once. However, this technique alone won't work for atomic nonces. The atomic ID is shared between parties, meaning a malicious attacker could attempt to trick a user into reusing an atomic ID.
+
+Currently, the automated defense is to use a space-efficient filter (`scalable cuckoo filter`), which is stored in the wallet's database. The filter cannot be reconstructed if the wallet file is deleted and recovered, so it is recommended to use a new keymask for any future atomic swaps.
+
+## Adaptor signatures
+[adaptor-signatures]: #adaptor-signatures
+
+Adaptor signatures were originally invented by Andrew Poelstra [\[3\]](#references). Adaptor signatures use additive properties of Schnorr signatures to commit to a secret, then later reveal it by creating a full Schnorr signature over the same message.
+
+
+## Protocol rounds
+[protocol-rounds]: #protocol-rounds
+
+In Grin, let `kB` be Bob's random kernel nonce, `rB` be Bob's random blinding factor, `nB` be Bob's atomic nonce. Let `kA` and `rA` be Alice's random kernel nonce and blinding factor.
+
```

Comment:
The picture shows multiple outputs and transactions. So it's not clear which kernel and blinding factor you're talking about. Could you please number or name the various outputs and txs in the diagram,  and number blinding factors and nonces accordingly? The two secrets are unique, so those need no further qualification other than owner.

Reply:
> Could you please number or name the various outputs and txs in the diagram, and number blinding factors and nonces accordingly?

Yeah, I can add some clarifying descriptions for what the diagram depicts. Will make it clear that it is only depicting 3 transactions, and some details about the state changes. The protocol I've implemented doesn't exactly follow Somsen's description of "Succinct Atomic Swaps", since we would need NRD kernels.